src/commands/remoteDebugFunctionApp.ts

```diff
@@ -0,0 +1,125 @@
+
+// tslint:disable-next-line:no-require-imports
+import WebSiteManagementClient = require('azure-arm-website');
+import { SiteConfigResource, StringDictionary, User } from 'azure-arm-website/lib/models';
+import { DeviceTokenCredentials } from 'ms-rest-azure';
+import * as opn from "opn";
+import * as portfinder from 'portfinder';
+import * as vscode from 'vscode';
+import { SiteWrapper } from 'vscode-azureappservice';
+import { AzureTreeDataProvider, IAzureNode } from 'vscode-azureextensionui';
+import { DebugProxy } from '../DebugProxy';
+import { DialogResponses } from '../DialogResponses';
+import { localize } from '../localize';
+import { FunctionAppTreeItem } from '../tree/FunctionAppTreeItem';
+import { nodeUtils } from '../utils/nodeUtils';
+
+export async function remoteDebugFunctionApp(outputChannel: vscode.OutputChannel, tree: AzureTreeDataProvider, node?: IAzureNode<FunctionAppTreeItem>): Promise<void> {
+    const confirmMsg: string = localize('azFunc.confirmRemoteDebug', 'The configurations of the selected app will be changed before debugging. Would you like to continue?');
+    const result: vscode.MessageItem | undefined = await vscode.window.showWarningMessage(confirmMsg, DialogResponses.yes, DialogResponses.seeMoreInfo, DialogResponses.cancel);
+    if (result === DialogResponses.cancel) {
+        return;
+    } else if (result === DialogResponses.seeMoreInfo) {
+        // tslint:disable-next-line:no-unsafe-any
+        opn('https://aka.ms/azfunc-remotedebug');
+        return;
+    }
+
+    if (!node) {
+        node = <IAzureNode<FunctionAppTreeItem>>await tree.showNodePicker(FunctionAppTreeItem.contextValue);
+    }
+    const client: WebSiteManagementClient = nodeUtils.getWebSiteClient(node);
+    const siteWrapper: SiteWrapper = node.treeItem.siteWrapper;
+    const portNumber: number = await portfinder.getPortPromise();
+    const publishProfile: User = await siteWrapper.getWebAppPublishCredential(client);
+    const accessToken: string = await acquireToken(node.credentials);
+    const debugProxy: DebugProxy = new DebugProxy(outputChannel, siteWrapper, portNumber, publishProfile, accessToken);
+
+    await vscode.window.withProgress({ location: vscode.ProgressLocation.Window }, async (p: vscode.Progress<{}>) => {
+        // tslint:disable-next-line:no-any
+        return new Promise(async (resolve: () => void, reject: (e: any) => void): Promise<void> => {
+            try {
+                await updateSiteConfig(siteWrapper, client, p);
+                await updateAppSettings(siteWrapper, client, p);
+
+                p.report({ message: 'starting debug proxy...' });
+                // tslint:disable-next-line:no-floating-promises
+                debugProxy.startProxy();
+                debugProxy.on('start', resolve);
+            } catch (error) {
+                reject(error);
+            }
+        });
+    });
+
+    debugProxy.on('error', (err: Error) => {
```

Comment:
Could an error occur before start finishes? In other words, should this `on('error'` stuff be setup _before_ you wait for the 'start' event?